src/ralph/backends/data/fs.py

```diff
@@ -0,0 +1,315 @@
+            operation_type (BaseOperationType or None): The mode of the write operation.
+                If operation_type is `CREATE` or `INDEX`, the target file is expected to
+                    be absent. If the target file exists a `FileExistsError` is raised.
+                If operation_type is `UPDATE`, the target file is overwritten.
+                If operation_type is `APPEND`, the data is appended to the
+                    end of the target file.
+
+        Returns:
+            int: The number of written files.
+
+        Raises:
+            BackendException: If the `operation_type` is `CREATE` or `INDEX` and the
+                target file already exists.
+            BackendParameterException: If the `operation_type` is `DELETED` as it is not
+                supported.
+        """
+        data = iter(data)
+        try:
+            first_record = next(data)
+        except StopIteration:
+            logger.info("Data Iterator is empty; skipping write to target.")
+            return 0
+        if not operation_type:
+            operation_type = self.default_operation_type
+
+        if operation_type == BaseOperationType.DELETE:
+            msg = "Delete operation_type is not allowed."
+            logger.error(msg)
+            raise BackendParameterException(msg)
+
+        if not target:
+            target = f"{now()}-{uuid4()}"
+            logger.info("Target file not specified; using random file name: %s", target)
+
+        target = Path(target)
+        path = target if target.is_absolute() else self.default_directory / target
+
+        if operation_type in [BaseOperationType.CREATE, BaseOperationType.INDEX]:
+            if path.is_file():
+                msg = (
+                    "%s already exists and overwrite is not allowed with operation_type"
+                    " create or index."
+                )
+                logger.error(msg, path)
+                raise BackendException(msg % path)
+
+            logger.debug("Creating file: %s", path)
+
+        mode = "wb"
+        if operation_type == BaseOperationType.APPEND:
+            mode = "ab"
+            logger.debug("Appending to file: %s", path)
+
+        with path.open(mode) as file:
+            is_dict = isinstance(first_record, dict)
+            writer = self._write_dict if is_dict else self._write_raw
+            for chunk in chain((first_record,), data):
+                writer(file, chunk)
+
+        # The file has been created, add a new entry to the history.
+        self.append_to_history(
+            {
+                "backend": self.name,
+                "action": "write",
+                "id": str(path),
+                "filename": path.name,
+                "size": path.stat().st_size,
+                "timestamp": now(),
+            }
+        )
+        return 1
+
+    @staticmethod
+    def _read_raw(file: IO, chunk_size: int, _ignore_errors: bool) -> Iterator[bytes]:
```

Comment:
Should we be so specific for the `file` type ?

Reply:
What would be a less specific type hint for a file-like object? We could use IOBase or Iterable in `_read_dict`, however in `_read_raw` we use the `read` method on the file which is part of the `IO` class.